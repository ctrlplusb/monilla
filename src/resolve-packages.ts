import globby from "globby";
import path from "path";
import { PackageJson } from "type-fest";

import { ErrorCode, YalcdError } from "./yalcd-error";

export type PackageMeta = {
  name: string;
  path: string;
  packageJson: PackageJson;
  isRoot: boolean;
};

export function resolvePackages(rootDir: string): PackageMeta[] {
  const packageJsons: PackageMeta[] = [];

  const rootPackageJsonPath = path.join(rootDir, "package.json");

  const packageJsonPaths = globby.sync(path.join(rootDir, "/**/package.json"));

  for (const packageJsonPath of packageJsonPaths) {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const packageJson: PackageJson = require(packageJsonPath);

    if (packageJson.name == null || packageJson.name.trim().length === 0) {
      throw new YalcdError(
        ErrorCode.InvalidPackageJSON,
        `Package "${packageJsonPath}" has no name.`,
      );
    }

    packageJsons.push({
      name: packageJson.name,
      path: packageJsonPath.replace(rootDir, "."),
      packageJson,
      isRoot: rootPackageJsonPath === packageJsonPath,
    });
  }

  return packageJsons;
}
