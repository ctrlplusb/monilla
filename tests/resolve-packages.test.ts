import path from "path";
import tempy from "tempy";

import { resolvePackages } from "~/resolve-packages";

describe("resolvePackages", () => {
  it("should return an empty array if no packages are found", () => {
    const actual = resolvePackages(tempy.directory());

    expect(actual).toEqual([]);
  });

  it("[integration] should resolve the expected packages", () => {
    const actual = resolvePackages(
      path.join(__dirname, "__fixtures__/example-monorepo"),
    );

    expect(actual).toMatchSnapshot();
  });
});
