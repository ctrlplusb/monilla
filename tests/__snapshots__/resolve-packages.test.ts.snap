// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolvePackages [integration] should resolve the expected packages 1`] = `
Array [
  Object {
    "isRoot": true,
    "name": "example-monorepo",
    "packageJson": Object {
      "description": "Example monorepo to aid testing of yalcd",
      "name": "example-monorepo",
      "private": true,
      "version": "1.0.0",
    },
    "path": "./package.json",
  },
  Object {
    "isRoot": false,
    "name": "@my/terminal",
    "packageJson": Object {
      "author": "Sean Matheson",
      "dependencies": Object {
        "@my/messages": "file:../messages",
      },
      "description": "",
      "keywords": Array [],
      "license": "ISC",
      "main": "index.js",
      "name": "@my/terminal",
      "version": "1.0.0",
    },
    "path": "./apps/terminal/package.json",
  },
  Object {
    "isRoot": false,
    "name": "@my/messages",
    "packageJson": Object {
      "main": "index.js",
      "name": "@my/messages",
      "private": true,
      "version": "1.0.0",
    },
    "path": "./packages/messages/package.json",
  },
]
`;
